File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
import sys

# 백트래킹
N, M = map(int, sys.stdin.readline().split())
s = []
visited = [False for _ in range(N + 1)]


def DFS():
    if len(s) == M:
        print(' '.join(map(str, s)))
        return
    for i in range(1, N + 1):
        if visited[i]:
            continue
        visited[i] = True
        s.append(i)
        DFS()  # 재귀로 값을 다 구하면
        visited[i] = False  # 원래 있던 상태로 변경
        s.pop()


DFS()

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.*;
import java.util.*;
public class Main{

    public static boolean[] visited;
    public static Stack<Integer> answer;
    public static StringBuilder sb;

    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        sb = new StringBuilder();

        Integer n = Integer.parseInt(st.nextToken());
        Integer m = Integer.parseInt(st.nextToken());

        visited = new boolean[n+1];
        answer = new Stack<>();
        calculatePer(0, n, m);

        System.out.println(sb);
    }

    public static void calculatePer(int depth, int n, int m){
        if(m == depth) {
            for (Integer integer : answer) {
                sb.append(integer).append(' ');
            }
            sb.append('\n');
        }

        for(int i = 1; i <= n; i ++){
            if(visited[i]) continue;
            visited[i] = true;
            answer.add(i);
            calculatePer(depth+1, n, m);
            visited[i] = false;
            answer.pop();
        }
    }


}